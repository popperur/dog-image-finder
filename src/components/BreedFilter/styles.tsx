import styled from "styled-components";
import { Button } from "antd";

export const ResetButton = styled(Button)`
  position: absolute;
  right: 20px;
`;
