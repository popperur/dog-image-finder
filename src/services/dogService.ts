import axios from "axios";
import capitalize from "antd/lib/_util/capitalize";
import handleError from "utils/errorHandler.ts";

const BASE_URL = "https://dog.ceo/api";

type DogBreeds = Record<string, string[]>;

// Maps dog.ceo breed names
// from {spaniel:["japanese","irish","cocker"], komondor: []}
// to ["Cocker Spaniel", "Irish Spaniel", "Japanese Spaniel", "Komondor"]
function transformBreedNames(dogBreeds: DogBreeds) {
  const breedNames: string[] = [];
  for (const breedName in dogBreeds) {
    const subBreeds = dogBreeds[breedName];
    if (subBreeds.length > 0) {
      subBreeds.forEach((subBreed: string) =>
        breedNames.push(`${capitalize(subBreed)} ${capitalize(breedName)}`),
      );
    } else {
      breedNames.push(capitalize(breedName));
    }
  }
  breedNames.sort((a, b) => a.localeCompare(b));
  return breedNames;
}

async function getBreedNames() {
  return axios
    .get(`${BASE_URL}/breeds/list/all`)
    .then(response => {
      if (response.data.status !== "success") {
        console.error(
          "Dog.ceo retrieved the message with a non-success status.",
        );
        return [];
      }
      return transformBreedNames(response.data.message);
    })
    .catch(error => handleError(error, []));
}

export { getBreedNames };
