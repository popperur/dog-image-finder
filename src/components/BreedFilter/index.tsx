import { useEffect, useRef, useState } from "react";
import { AutoComplete } from "antd";
import { CloseOutlined } from "@ant-design/icons";
import { getBreedNames } from "services/dogService.ts";
import { BaseSelectRef } from "rc-select";
import { ResetButton } from "components/BreedFilter/styles.tsx";
import { isMobileDevice } from "utils/mobile.ts";

interface BreedFilterProps {
  selectedBreedName: string;
  onBreedNameSelect: (breedName: string) => void;
}

function BreedFilter({
  selectedBreedName,
  onBreedNameSelect,
}: BreedFilterProps) {
  const [inputValue, setInputValue] = useState<string>("");
  const [breedNames, setBreedNames] = useState<string[]>([]);
  const [filteredBreedNames, setFilteredBreedNames] = useState<string[]>([]);
  const [dropdownVisible, setDropdownVisible] = useState<boolean>(false);
  const autoCompleteRef = useRef<BaseSelectRef>(null);

  useEffect(() => {
    getBreedNames()
      .then(breedNames => {
        setBreedNames(breedNames);
        setFilteredBreedNames(breedNames);
      })
      .catch(error => {
        console.error(`Error in BreedFilter: ${error}`);
      });
  }, []);

  const handleSearch = (typedText: string) => {
    const blank = typedText === "";
    setDropdownVisible(!blank);
    if (!blank) {
      const filteredNames: string[] = breedNames.filter(breedName =>
        breedName.toLowerCase().includes(typedText.toLowerCase()),
      );
      setFilteredBreedNames(filteredNames);
    }
  };

  const handleSelect = (breedName: string) => {
    onBreedNameSelect(breedName);
    setDropdownVisible(false);
    if (isMobileDevice()) {
      autoCompleteRef.current?.blur();
    }
  };

  const onChange = (data: string) => {
    setInputValue(data);
  };

  const resetSelection = () => {
    setInputValue("");
    handleSelect("");
    autoCompleteRef.current?.focus();
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (
      (e.key === "ArrowDown" || e.key === "Enter") &&
      inputValue === "" &&
      !dropdownVisible
    ) {
      setFilteredBreedNames(breedNames);
      setDropdownVisible(true);
    }
  };

  return (
    <>
      <AutoComplete
        aria-label="breed name autocomplete"
        ref={autoCompleteRef}
        value={inputValue}
        options={filteredBreedNames.map(breedName => ({ value: breedName }))}
        style={{ width: 260 }}
        onSelect={handleSelect}
        onSearch={handleSearch}
        onChange={onChange}
        onKeyDown={handleKeyDown}
        open={dropdownVisible}
        placeholder="Look up a good boy by breed name!"
      />

      {selectedBreedName && (
        <ResetButton
          aria-label="reset selection"
          color="default"
          shape="default"
          icon={<CloseOutlined />}
          onClick={resetSelection}
        />
      )}
    </>
  );
}

export default BreedFilter;
