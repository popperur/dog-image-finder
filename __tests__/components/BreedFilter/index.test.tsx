import { beforeEach, describe, expect, vi } from "vitest";
import BreedFilter from "components/BreedFilter";
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { getBreedNames } from "services/dogService.ts";

describe("BreedFilter component", () => {
  describe("Snapshot", () => {
    it("matches the snapshot", async () => {
      const { asFragment } = render(
        <BreedFilter selectedBreedName="" onBreedNameSelect={vi.fn()} />,
      );

      await waitFor(() => {
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe("Breed name selector", () => {
    beforeEach(() => {
      vi.mock("services/dogService", () => ({
        getBreedNames: vi.fn().mockResolvedValue(["Komondor", "Terrier"]),
      }));
    });

    it("fetches the breed names", async () => {
      render(<BreedFilter selectedBreedName="" onBreedNameSelect={vi.fn()} />);

      await waitFor(() => {
        expect(getBreedNames).toHaveBeenCalled();
      });
    });

    it("shows the breed names in the dropdown", async () => {
      render(<BreedFilter selectedBreedName="" onBreedNameSelect={vi.fn()} />);

      fireEvent.change(screen.getByRole("combobox"), {
        target: { value: "Kom" }, // Simulating user typing
      });

      const komondorOption = await screen.findByTitle("Komondor");
      expect(komondorOption).toBeInTheDocument();
    });

    it("logs the selected breed", async () => {
      render(<BreedFilter selectedBreedName="" onBreedNameSelect={vi.fn()} />);

      const consoleLogMock = vi
        .spyOn(console, "log")
        .mockImplementation(() => {});

      fireEvent.change(screen.getByRole("combobox"), {
        target: { value: "Kom" },
      });

      const komondorOption = await screen.findByTitle("Komondor");
      fireEvent.click(komondorOption);

      expect(consoleLogMock).toHaveBeenCalledWith(
        "Selected breed name: Komondor",
      );

      consoleLogMock.mockRestore();
    });
  });

  describe("Reset button", () => {
    it("is rendered if there is a selection", async () => {
      render(
        <BreedFilter
          selectedBreedName="Komondor"
          onBreedNameSelect={vi.fn()}
        />,
      );

      await waitFor(() => {
        const resetButton = screen.getByRole("button", { name: /reset selection/i });
        expect(resetButton).toBeInTheDocument();
      });
    });

    it("is not rendered if there is no selection", async () => {
      render(<BreedFilter selectedBreedName="" onBreedNameSelect={vi.fn()} />);

      await waitFor(() => {
        const resetButton = screen.queryByRole("button", { name: /reset selection/i });
        expect(resetButton).not.toBeInTheDocument();
      });
    });

    it("clears the selection", async () => {
      const mockOnBreedNameSelect = vi.fn();

      render(
        <BreedFilter
          selectedBreedName="Komondor"
          onBreedNameSelect={mockOnBreedNameSelect}
        />,
      );

      await waitFor(() => {
        const resetButton = screen.getByRole("button", { name: /reset selection/i });
        fireEvent.click(resetButton);
        expect(mockOnBreedNameSelect).toHaveBeenCalledWith("");
      });
    });
  });
});
