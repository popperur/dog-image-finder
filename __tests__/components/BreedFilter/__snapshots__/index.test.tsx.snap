// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreedFilter component > snapshot > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="breed name autocomplete"
    class="ant-select ant-select-outlined ant-select-auto-complete css-dev-only-do-not-override-1r287do ant-select-single ant-select-show-search"
    style="width: 260px;"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <input
          aria-autocomplete="list"
          aria-controls="rc_select_TEST_OR_SSR_list"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="breed name autocomplete"
          aria-owns="rc_select_TEST_OR_SSR_list"
          autocomplete="off"
          class="ant-select-selection-search-input"
          id="rc_select_TEST_OR_SSR"
          role="combobox"
          type="search"
          value=""
        />
      </span>
      <span
        class="ant-select-selection-placeholder"
      >
        Look up a good boy by breed name!
      </span>
    </div>
  </div>
</DocumentFragment>
`;
